File: Algorithms/Data Structure dependent/Trees/BST/BSTIterator.cpp
class BSTIterator {
public:
    BSTIterator(TreeNode* root) {
        cur = root;
    }

    int next() {
        assert(hasNext());
        auto nextNode = getNext();
        return nextNode->val;
    }
    
    bool hasNext() {
        return !stack.empty() || cur;
    }
    
private:
    TreeNode* cur;
    std::stack<TreeNode*> stack;

    TreeNode* getNext(){
        /* Go left as far as we can */
        while(cur != nullptr){
            stack.push(cur);
            cur = cur->left;
        }
        /* After we process current node - we try to go right subtree */
        TreeNode* next = stack.top(); stack.pop();
        cur = next->right;

        return next;
    }
};

File: Algorithms/Data Structure dependent/Trees/BST/BSTIterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:
    cur = None
    stack = []

    def __init__(self, root: Optional[TreeNode]):
        self.cur = root
        self.stack = []

    def next(self) -> int:
        self.pushLeft()
        nextNode = self.stack.pop()
        self.cur = nextNode.right

        return nextNode.val

    def hasNext(self) -> bool:
        return self.cur or len(self.stack)

    def pushLeft(self):
        root = self.cur 
        while root:
            self.stack.append(root)
            root = root.left

